Parse push request after distributor, ingester validate limits

dimitarvdimitrov opened the pull request:
Signed-off-by: Dimitar Dimitrov <dimitar.dimitrov@grafana.com>

<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Rebase your PR if it gets out of sync with main
-->

#### What this PR does

The ingester and distributor push functions have some  limit validations
 on push requests that do not depend on the push request itself (e.g.
 request rate or inflight requests). This PR makes parsing of request
 happen only after the distributor and ingester have validated as many
 of their instance limits as possible. This should reduce utilization
 when the distributor and ingesters are under load.

Originally the issue was for the distributor only, but since their `PushWithCleanup` methods both confirm to the `push.Func` signature, I decided it is easier to change both of them.

I am submitting this PR as a draft because it complicates the code,
and only helps when requests in the distributor start taking a very long time and when effectively all incoming requests get counted as over the inflight limit. So I am not sure if this is something we want to include.

#### Which issue(s) this PR fixes or relates to

Fixes https://github.com/grafana/mimir/issues/2327

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pstibrany commented:
What if distributor exposed new handler with the checks, that we would use in front of push handler?

For example:

```
func (d *Distributor) CheckLimitsHandler(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// We will report *this* request in the error too.
		inflight := d.inflightPushRequests.Inc()

		if d.cfg.InstanceLimits.MaxInflightPushRequests > 0 && inflight > int64(d.cfg.InstanceLimits.MaxInflightPushRequests) {
			http.Error(w, errMaxInflightRequestsReached.Error(), http.StatusInternalServerError)
			return
		}

		defer func() {
			d.inflightPushRequests.Dec()
		}()

		next.ServeHTTP(w, r)
	})
}
```

And then we would register this in `api.go`:

```go
func (a *API) RegisterDistributor(d *distributor.Distributor, pushConfig distributor.Config) {
	...
	a.RegisterRoute("/api/v1/push", d.CheckLimitsHandler(push.Handler(pushConfig.MaxRecvMsgSize, a.sourceIPs, a.cfg.SkipLabelNameValidationHeader, wrappedDistributor)), true, false, "POST")
	...
}
```

dimitarvdimitrov commented:
That indeed makes the code simpler. Right now my problem is that `Push` and `PushWithCleanup` have different semantics - one checks some limits, the other doesn't. I tried explaining that in the godoc. Do you think I should be worrying about this? It's the semantics of the exported methods from the distributor, but the distributor is also quite top-level, so it doesn't have many clients. The clients it has, use the grpc interface (which retains the same behaviour as before).

pstibrany commented on pkg/distributor/distributor.go:
```diff
@@ -595,22 +596,78 @@ func (d *Distributor) forwardingReq(ctx context.Context, userID string) forwardi
+	return nil
+}
+
+// Push implements client.IngesterServer
```
```suggestion
// Push is gRPC method registered as client.IngesterServer and distributor.DistributorServer.
```

pstibrany commented on pkg/distributor/distributor.go:
```diff
@@ -595,22 +596,78 @@ func (d *Distributor) forwardingReq(ctx context.Context, userID string) forwardi
-// Push implements client.IngesterServer
-func (d *Distributor) Push(ctx context.Context, req *mimirpb.WriteRequest) (*mimirpb.WriteResponse, error) {
-	return d.PushWithCleanup(ctx, req, func() { mimirpb.ReuseSlice(req.Timeseries) })
+func (d *Distributor) PrePushLimitsHandler(next http.Handler) http.Handler {
```
```suggestion
// PrePushLimitsHandler is used as HTTP middleware in front of PushWithCleanup method.
func (d *Distributor) PrePushLimitsHandler(next http.Handler) http.Handler {
```

pstibrany commented on pkg/distributor/distributor.go:
```diff
@@ -595,22 +596,78 @@ func (d *Distributor) forwardingReq(ctx context.Context, userID string) forwardi
+
+// PushWithCleanup takes a WriteRequest and distributes it to ingesters using the ring.
+// Strings in `req` may be pointers into the gRPC buffer which will be reused, so must be copied if retained.
+// PushWithCleanup does not check limits like ingestion rate and inflight requests. Use Push to check those.
```
```suggestion
// PushWithCleanup does not check limits like ingestion rate and inflight requests.
// These limits are checked either by Push gRPC method (when invoked via gRPC) or PrePushLimitsHandler (when invoked via HTTP)
```

pstibrany commented on pkg/api/api.go:
```diff
@@ -241,7 +241,7 @@ func (a *API) RegisterDistributor(d *distributor.Distributor, pushConfig distrib
 	wrappedDistributor := a.cfg.wrapDistributorPush(d)
 
-	a.RegisterRoute("/api/v1/push", push.Handler(pushConfig.MaxRecvMsgSize, a.sourceIPs, a.cfg.SkipLabelNameValidationHeader, wrappedDistributor), true, false, "POST")
+	a.RegisterRoute("/api/v1/push", d.PrePushLimitsHandler(push.Handler(pushConfig.MaxRecvMsgSize, a.sourceIPs, a.cfg.SkipLabelNameValidationHeader, wrappedDistributor)), true, false, "POST")
```
Shall we use pre-push limits handler for OLTP handler too?

pstibrany reviewed (commented):
👍

pstibrany commented:
> But now I introduced another problem - `Push` and `PushWithCleanup` have different semantics - one checks some limits, the other doesn't. I tried explaining that in the godoc. Do you think I should be worrying about this?

It is somewhat confusing, but that comes from the fact that `Push` is used by `gRPC` server directly, while `PushWithCleanup` is used from HTTP handler, so HTTP limit checking middleware applies to it. I don't see a simple way around it (gRPC middleware comes to mind, but it wouldn't be simple). May be finding better name for `PushWithCleanup` would help? (Not that I have any good ideas)

pstibrany reviewed (commented):
Changes look good to me. I think we should have a test to check that `PrePushLimitsHandler` works as expected. Distributor tests only verify the checks via `Push` code path. WDYT?

dimitarvdimitrov commented:
> Changes look good to me. I think we should have a test to check that `PrePushLimitsHandler` works as expected. Distributor tests only verify the checks via `Push` code path. WDYT?

agree, will add tests 


I realized that before this PR the behaviour was that the number of inflight requests was decremented once the last request to an ingester completed. After this PR, this behaviour will be different: the count will be decremented when we return an HTTP response to the request, while there may still be some inflight client requests to an ingester.

The behaviour was introduced with #398 and was intentional. I think we should retain it in this PR.

One way to do it is to make the middleware approach possible for `push.Func`s. This means changing their signature to accept something that supports lazy loading of the body. Then the `PrePushLimitsHandler` will be a `push.Func` instead of an `http.Handler` and will be able to pass a `cleanup` function that decrements the inflight requests. As for `push.Func`, I initially went with a function that returns the body, but I think if we change `req *mimirpb.WriteRequest` to a struct that has a singleflight method `GetRequest() *mimirpb.WriteRequest` we can get a better interface that will support a middleware approach. Will get back to this later, maybe I can thin of something else in the meantime.

pstibrany commented:
There is similar discussion in https://github.com/grafana/mimir/pull/2295.

dimitarvdimitrov commented:
Pushed some example middleware on a different branch just to demonstrate what I meant 435daaf6b23971f6ce8915ea9266d13492b60a14 It might not compile, also not sure where to put the ctx yet (commented the same on the wrong PR 🤦 )

dimitarvdimitrov commented:
so with the introduced middleware, the tests in the distributor also test the limits that the `PrePushLImitsHandler` (now `PrePushLimitsMiddleware`) impose.

I've also added a test that checks for what #398 added. PTAL.

dimitarvdimitrov commented on pkg/util/push/request.go:
```diff
@@ -0,0 +1,84 @@
+	var req mimirpb.PreallocWriteRequest
+	buf, err := r.parser(ctx, r.httpReq, r.maxMessageSize, bufHolder.buf, &req)
+	if err != nil {
+		//level.Error(logger).Log("err", err.Error())
+		//http.Error(w, err.Error(), http.StatusBadRequest)
```
I had forgotten about this and now realizing that this is introducing a bug - these errors will likely be reported as a 5xx by the `push.handler`. I'm working on a fix...

dimitarvdimitrov replied:
should be sorted in 3992f7bba6adac35cf756038fde29dd4884c4485

replay commented on pkg/util/push/request.go:
```diff
@@ -0,0 +1,84 @@
+		return
+	}
+	// If decoding allocated a bigger buffer, put that one back in the pool.
+	if len(buf) > len(bufHolder.buf) {
```
`bufHolder.buf` is a slice and not a pointer to a slice so its length property can't get updated by `r.parser` and at this point must still be set to whatever it was when it came out of the pool. For what you're trying to check here, wouldn't it be better to compare the `cap` instead of the `len`?

replay replied:
I just saw that `snappy.Decode` (used by `util.ParseProtoReader`) actually seems to look at the length of the given destination slice instead of the capacity, which I think is strange. Maybe we it would be better if we'd set `bufHolder.buf = bufHolder.buf[:cap(bufHolder.buf)]` before passing it into `r.parser`? That still feels strange to me though.

dimitarvdimitrov replied:
I noticed that, but snappy never returns a slice that has smaller length than capacity _and_ is not the same slice that was passed to it. If snappy returns a different slice it is because the original slice was too small and it allocated a bigger one. That's how it's been working so far AFAICT.

The capacity reset makes sense to me as a precaution. It's slightly out of scope of this PR, but I will add it

dimitarvdimitrov replied:
So I pushed 0192f40ab1dc5ddf018f8e1ae90968570a1fffcd which resets the capacity once it comes back from the parser. This way when you get a buffer from the pool you know that it `len(buf) == cap(buf)`. It has either been created via the bufferPool's factory which uses `make([]byte, N)` or has been maxed out by `buf = buf[:cap(buf)]`

replay replied:
Thanks!

dimitarvdimitrov commented:
I rebased and the PR should be ready for review again. Some of the things it was doing were already done by #2709.

One difference between 2709 and this PR is that 2709 leaves the per-tenant request rate limit check in the main `PushWithCleanup` method. I moved that into the `instanceLimitsMiddleware` and renamed the middleware to `limitsMiddleware`.

Also renamed `PushWithCleanup` to just `push` because it doesn't need to be exported and `WithCleanup` doesn't make sense anymore.

dimitarvdimitrov commented:
I am unsure if the rebase overhead is too much for the added benefit.

The benefit obviously depends on the load, but in an internal incident we found that the requests over the inflight request limit got rejected in <20ms. This means that we need a rate of 50 rejected requests per second (RRPS) to have, on average, a single in-memory request that will be later rejected because of the inflight requests limit. So for 10 in-memory requests we are at 500 RRPS. This is hard to come to in the real world especially because we have many distributors for a single cluster, so the load on the whole cluster needs to increase `* $number_of_distributors` to get a single in-memory request that will later be rejected.

This is still a DoS attack vector though.

@pstibrany @replay do you think this is still worth finishing or should I close it?

pstibrany commented on pkg/util/push/request.go:
```diff
@@ -0,0 +1,90 @@
+)
+
+type Request struct {
+	cleanups []func()
```
To avoid resizing this slice when adding cleanups, we could either preallocate it, or perhaps even use a big enough array (then we need extra field to know how many cleanup functions we have).

dimitarvdimitrov replied:
we currently call add cleanup up to 4 times in the distributor. Preallocated it with size 10

pstibrany commented on pkg/util/push/push.go:
```diff
@@ -67,43 +67,15 @@ func handler(maxRecvMsgSize int,
-		cleanup := func() {
-			mimirpb.ReuseSlice(req.Timeseries)
-			bufferPool.Put(bufHolder)
+		req := &Request{
+			cleanups:                     nil,
+			httpReq:                      r,
+			maxMessageSize:               maxRecvMsgSize,
+			allowSkipLabelNameValidation: allowSkipLabelNameValidation,
+			parser:                       parser,
 		}
```
Let's make use a `NewRequest(generatorFunction)` here. (`httpReq`, `maxMessageSize` and `allowSkipLabelNameValidation` will be handled by generator function)

pstibrany reviewed (commented):
I think we should get this through the finish line. I'd like to see `*Request` simplified, see my comments, but overall I think this is a good approach and this is pretty straight forward change now.

replay commented:
> I think we should get this through the finish line. 

I agree that the change makes sense, once Peter's comments are addressed

replay commented on pkg/util/push/push.go:
```diff
@@ -67,43 +67,15 @@ func handler(maxRecvMsgSize int,
-		cleanup := func() {
-			mimirpb.ReuseSlice(req.Timeseries)
-			bufferPool.Put(bufHolder)
+		req := &Request{
```
would it make sense to pool these `Request`s?

pstibrany replied:
I don't think we should do that without any profiling first.

dimitarvdimitrov replied:
@replay are you ok if I skip this from this PR?

replay replied:
Sure I'm ok with that, I meant it like a genuine question :)

dimitarvdimitrov commented on pkg/util/push/request.go:
```diff
@@ -0,0 +1,90 @@
+	maxMessageSize               int
+	allowSkipLabelNameValidation bool
+	parser                       ParserFunc
+	parseRequestOnce             sync.Once
```
that's true, that was a leftover from a previous version

dimitarvdimitrov commented on pkg/util/push/request.go:
```diff
@@ -0,0 +1,90 @@
+
+// CleanUp calls all added cleanups.
+func (r *Request) CleanUp() {
+	for _, f := range r.cleanups {
```
I could only find this which changes the cleanup and it invokes the previous cleanups before doing its own cleanup

https://github.com/grafana/mimir/blob/3ecb9dcf76a7c6976f022d8a05e983ac65d48aae/pkg/distributor/distributor.go#L855-L859

dimitarvdimitrov replied:
I am not sure which is a better sematic. Maybe calling the first cleanups last mimics defer semantics, so we want to be like that? But not sure what to do with the case ☝️ where we previously did the opposite.

pracucci replied:
Even if today the order doesn't matter, I think it would be a preferred behaviour treating the cleanup function as a stack, so calling them in reverse order (looks more future proof to me).

dimitarvdimitrov replied:
I pushed a change to do stack semantics in https://github.com/grafana/mimir/commit/e6ad4508ffa5260384eab490297bb39bc05db44d

dimitarvdimitrov commented on pkg/util/push/request.go:
```diff
@@ -0,0 +1,90 @@
+
+// WriteRequest parses the body of the push request and caches it.
+// Subsequent calls to WriteRequest return the same results.
+func (r *Request) WriteRequest(ctx context.Context) (*mimirpb.WriteRequest, error) {
```
`ctx` is used while parsing the body, in case the ctx is cancelled while parsing is taking a long time. Later invocations to `WriteRequest` don't respect context cancellations because they return immediately.

dimitarvdimitrov commented on pkg/util/push/request.go:
```diff
@@ -0,0 +1,90 @@
+	parseError                   error
+}
+
+func (r *Request) parseRequest(ctx context.Context) {
```
I extracted some of it back into `push.handler` in  https://github.com/grafana/mimir/pull/2419/commits/8e4eba3fd3483fb1cb1a2d6309dd03a05ca27337

dimitarvdimitrov replied:
and then a few more changes later, you should check the general diff again

pstibrany reviewed (approved):
Good job, thank you! I left some small comments, but nothing blocking.

dimitarvdimitrov commented on pkg/util/push/request.go:
```diff
@@ -0,0 +1,64 @@
+// WriteRequest parses the body of the push request and caches it.
+// Subsequent calls to WriteRequest return the same results.
+func (r *Request) WriteRequest() (*mimirpb.WriteRequest, error) {
+	if !r.requestIsParsed {
```
it kind of changes the contract with the supplier function. Now it will be invoked only once if it returns a non-nil error or a body. Otherwise it will be invoked on every `WriteRequest()` call. I would prefer to keep it as-is

pstibrany replied:
> Otherwise

This means it returned `nil` request and error. I'd say that's an error :)

dimitarvdimitrov replied:
I'm not sure about this one. The error is returned on an exported method and mentions some package-internal implementation details. I've pushed a change in 1292ded544cbc14ffb0c8a37a56aa1548186b13c which does this. If you're happy with it, i'm ok to keep it.

replay commented on pkg/util/push/push.go:
```diff
@@ -85,43 +84,38 @@ func handler(maxRecvMsgSize int,
+		supplier := func() (*mimirpb.WriteRequest, func(), error) {
+			bufHolder := bufferPool.Get().(*bufHolder)
+			var req mimirpb.PreallocWriteRequest
+			buf, err := parser(ctx, r, maxRecvMsgSize, bufHolder.buf, &req)
+			if err != nil {
+				// Check for httpgrpc error, default to client error if parsing failed
+				if _, ok := httpgrpc.HTTPResponseFromError(err); !ok {
+					err = httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+				}
+
+				bufferPool.Put(bufHolder)
+				return nil, nil, err
+			}
+			// If decoding allocated a bigger buffer, put that one back in the pool.
+			if buf = buf[:cap(buf)]; len(buf) > len(bufHolder.buf) {
+				bufHolder.buf = buf
 			}
```
would this not be the same and more similar to similar to other typical usages of slices, like for example `buf = append(buf, ...)`?

```suggestion
			var err error
			bufHolder.buf, err = parser(ctx, r, maxRecvMsgSize, bufHolder.buf, &req)
			bufHolder.buf = bufHolder.buf[:cap(bufHolder.buf)]
			if err != nil {
				// Check for httpgrpc error, default to client error if parsing failed
				if _, ok := httpgrpc.HTTPResponseFromError(err); !ok {
					err = httpgrpc.Errorf(http.StatusBadRequest, err.Error())
				}

				bufferPool.Put(bufHolder)
				return nil, nil, err
			}
```

dimitarvdimitrov replied:
so I wrote this 6c8ea108bca570bd071be94d3e325a0fde45f5b1, but it has the caveat that if the returned `buf` is indeed a different slice, then we retain both slices until the cleanup is called - at the end of the request - instead of during the request parsing.

dimitarvdimitrov replied:
because of this I reverted the commit in https://github.com/grafana/mimir/pull/2419/commits/ceaab27804030e92eefc940c96f8f6dfaa1ee584

replay reviewed (approved):
Looks great, I only added some nits

dimitarvdimitrov commented:
I will rebase this soon and try to get it in a mergeable condition

pracucci commented on pkg/util/push/request.go:
```diff
@@ -0,0 +1,65 @@
+type supplierFunc func() (*mimirpb.WriteRequest, func(), error)
+
+// Request represents a push request. It allows lazy body reading from the underlying http request
+// and adding cleanups that should be done after the request is completed.
```
[nit]

```suggestion
// and adding cleanup functions that should be called after the request has been handled.
```

pracucci commented on pkg/util/push/request.go:
```diff
@@ -0,0 +1,65 @@
+
+func newRequest(p supplierFunc) *Request {
+	return &Request{
+		cleanups:   make([]func(), 0, 10),
```
10 looks many cleanups. How many do we actually set in the worst case scenario?

dimitarvdimitrov replied:
that was to future-proof it. 4 at the moment

pstibrany replied:
Btw, we can avoid one allocation here by using 1) array, 2) slice that array:
```
type struct Request {
  cleanupsArr [10]func()
  cleanups []func()
...
}

r.cleanups: cleanupsArr[:0]
```

That makes request itself bigger, but no extra allocation for backing array is required.

dimitarvdimitrov replied:
neat. I pushed the change in https://github.com/grafana/mimir/pull/2419/commits/a0b383b7b1dc6bad34f992a15b85b5853e7c29ed

pracucci commented on pkg/util/push/request.go:
```diff
@@ -0,0 +1,65 @@
+	"github.com/grafana/mimir/pkg/mimirpb"
+)
+
+// supplierFunc should return either a non-nil body or a non-nil error. The cleanup function can be nil.
```
[nit]

```suggestion
// supplierFunc should return either a non-nil body or a non-nil error. The returned cleanup function can be nil.
```

pracucci commented on pkg/util/push/request.go:
```diff
@@ -0,0 +1,65 @@
+)
+
+// supplierFunc should return either a non-nil body or a non-nil error. The cleanup function can be nil.
+type supplierFunc func() (*mimirpb.WriteRequest, func(), error)
```
[nit] I would name the return params for clarity.

```suggestion
type supplierFunc func() (req *mimirpb.WriteRequest, cleanup func(), err error)
```

pracucci commented on pkg/util/push/push.go:
```diff
@@ -85,43 +84,38 @@ func handler(maxRecvMsgSize int,
+			var req mimirpb.PreallocWriteRequest
+			buf, err := parser(ctx, r, maxRecvMsgSize, bufHolder.buf, &req)
+			if err != nil {
+				// Check for httpgrpc error, default to client error if parsing failed
```
This looks like a logic change to me. Before we were guaranteeing to always return a HTTP error, while now we could return a httpgrpc error. Isn't a regression?

dimitarvdimitrov replied:
that's true, I had conflated the httpgrpc and http packages. It also means we also need a test for this, no?

dimitarvdimitrov replied:
part of the translation was moved elsewhere actually 

https://github.com/grafana/mimir/blob/bd40d46d9c8991a2fef7f529f35674eb6db33127/pkg/util/push/push.go#L119-L132

so what you're commenting on is the error that will be returned to whoever did `_, err := push.Request.WriteRequest()`. This error wasn't previously exposed outside the push package, so it's new behaviour.

I added more tests in 70d8de261d7b7faa0d963043b8271f1cb78f95f3

pracucci replied:
You're right. LGTM, thanks.

pracucci commented on CHANGELOG.md:
```diff
@@ -138,6 +138,7 @@
 * [ENHANCEMENT] Alertmanager: reduced memory utilization in Mimir clusters with a large number of tenants. #3143
 * [ENHANCEMENT] Store-gateway: added extra span logging to improve observability. #3131
 * [ENHANCEMENT] Compactor: cleaning up different tenants' old blocks and updating bucket indexes is now more independent. This prevents a single tenant from delaying cleanup for other tenants. #2631
+* [ENHANCEMENT] Distributor: most limits are now enforce before reading and parsing the body of the request. This makes the distributor more resilient against a burst of invalid requests or requests over the rate limit. #2419
```
> most limits

Please clarify which limits.

dimitarvdimitrov replied:
https://github.com/grafana/mimir/pull/2419/commits/0456dfd8cdaa6cbf090b36d26574aeda134f2524

pracucci reviewed (approved):
LGTM, thanks!
